### PR TITLE
Account for subresources and subpaths in nonresources

### LINKED ISSUE #590: Authorizer doesn't regard subresource from policy
/kind bug

**What happened**:
Whenever policy has resource entries with subresources, the never get matched. 
This is because:
1. subresource is not present in authorizer.AttributesRecord that is built from 
resourceAttributes;
2. resourceMatches function doesn't check for subresources;

**What you expected to happen**:
Expect policy to match in case where "resources" field in policy has records like
"pods/exec"

**How to reproduce it (as minimally and precisely as possible)**:

1. Create policy containing record similar to show below
`{ 
        "resource": {
          "verbs": ["create"],
          "resources": ["pods/exec"],
          "version": "*",
          "namespace": "utility"
        },
        "match": [
          {
            "type": "role",
            "values": ["utility_exec"]
          }
        ]
      },
` 
2. Try to execute to pod accordingly to policy. Make sure that nonresource entry doesn't allow all verbs on all resources otherwise it may get matched instead of actual resource 

**Anything else we need to know?**:

**Environment**:
- openstack-cloud-controller-manager version: all versions
- OS (e.g. from /etc/os-release): "16.04.4 LTS (Xenial Xerus)"
- Kernel (e.g. `uname -a`): Linux 4.15.0-34-generic #37~16.04.1-Ubuntu SMP Tue Aug 28 10:44:06 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

### PR DESCRIPTION
**Pull request by avlasov-mos-de**
**What this PR does / why we need it**:
Make sure AttributesRecord sturct gets Subresource field
updated with corresponding subresource. This is required
when accessing k8s resources such as "pods/exec", "pods/log".
Account for subresource field when check for match with policy
is done.
Account for subpath match during check of nonresource paths.
This is required as during resouece queries additional queries
to path such as "/api/*" or /apis/* are made.
Add tests to verify mentioned fucntionality.
Update policy.json file used during tests.
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #590

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by ramineni** [comment]
@avlasov-mos-de Thanks for the PR. Please update PR description on what is the PR about and also you can link issue number in fixes #590 in PR description only. And also, you need to get the cla signed.https://github.com/kubernetes/cloud-provider-openstack/pull/591#issuecomment-483911236

**Comment by avlasov-mos-de** [comment]
thanks @ramineni, I'll do that. Also I noticed that I'd missed something in my PR, I'll work on fixing the issue. I guess I also will need to update tests and test policy file to reflect subresurces.

**Comment by avlasov-mos-de** [comment]
Hi @ramineni and @reviewers, l've updated PR with what was asked. Looks like two tests that are failing are expected to fail. Please let me know if there is something else required on mi side

**Comment by lingxiankong** [comment]
I have a question to the issue, please see my comment in #590

**Comment by lingxiankong** [comment]
Changes look good to me, could you please add a new section under `Prepare the authorization policy` in the doc <https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/using-keystone-webhook-authenticator-and-authorizer.md>?

**Comment by avlasov-mos-de** [comment]
Sure, will do.

File: docs/using-keystone-webhook-authenticator-and-authorizer.md [hunk 1]
```diff
@@ -109,6 +108,104 @@ $ kubectl apply -f /etc/kubernetes/keystone-auth/policy-config.yaml
 As you can see, the version 2 policy definition is much simpler and
 more succinct.
 
+For many scenarios clients require access to `nonresourse` paths.
+`nonresource` paths include: `/api`, `/apis`, `/metrics`, `/resetMetrics`,
+`/logs`, `/debug`, `/healthz`, `/swagger-ui/`, `/swaggerapi/`, `/ui`, and
+`/version`. Clients require access to `/api`, `/api/*`, `/apis`, `/apis/*`,
+and /version to discover what resources and versions are present on the
```
Comment (lingxiankong) [inline docs/using-keystone-webhook-authenticator-and-authorizer.md]: /version -> `/version`

File: docs/using-keystone-webhook-authenticator-and-authorizer.md [hunk 2]
```diff
@@ -109,6 +108,104 @@ $ kubectl apply -f /etc/kubernetes/keystone-auth/policy-config.yaml
...
+server. Access to other `nonresource` paths can be disallowed without
+restricting access to the REST api.
+
```
Comment (lingxiankong) [inline docs/using-keystone-webhook-authenticator-and-authorizer.md]: I'd prefer to put the above paragraph into a sub-section named `Non-resource permission` under `Prepare the authorization policy` section, and all the following new content into another sub-section named `Sub-resource permission`, what do you think?

File: docs/using-keystone-webhook-authenticator-and-authorizer.md [hunk 3]
```diff
@@ -109,6 +108,104 @@ $ kubectl apply -f /etc/kubernetes/keystone-auth/policy-config.yaml
 As you can see, the version 2 policy definition is much simpler and
 more succinct.
 
+For many scenarios clients require access to `nonresourse` paths.
+`nonresource` paths include: `/api`, `/apis`, `/metrics`, `/resetMetrics`,
+`/logs`, `/debug`, `/healthz`, `/swagger-ui/`, `/swaggerapi/`, `/ui`, and
+`/version`. Clients require access to `/api`, `/api/*`, `/apis`, `/apis/*`,
+and /version to discover what resources and versions are present on the
+server. Access to other `nonresource` paths can be disallowed without
+restricting access to the REST api.
```
Comment (lingxiankong) [inline docs/using-keystone-webhook-authenticator-and-authorizer.md]: > Access to other `nonresource` paths can be disallowed without restricting access to the REST api.

I'm a little confused about this line, what do you want to describe?

File: docs/using-keystone-webhook-authenticator-and-authorizer.md [hunk 4]
```diff
@@ -109,6 +108,104 @@ $ kubectl apply -f /etc/kubernetes/keystone-auth/policy-config.yaml
...
+
+In order to describe subresource (e.g `logs` or `exec`) of a certain resource
+(e.g. `pod`)it is possible to us `/` in order to combine resource and
```
Comment (lingxiankong) [inline docs/using-keystone-webhook-authenticator-and-authorizer.md]: us -> use

File: docs/using-keystone-webhook-authenticator-and-authorizer.md (hunk 2 above)
Comment (avlasov-mos-de) [reply docs/using-keystone-webhook-authenticator-and-authorizer.md]: yes, probably would make more sense

File: docs/using-keystone-webhook-authenticator-and-authorizer.md (hunk 3 above)
Comment (avlasov-mos-de) [reply docs/using-keystone-webhook-authenticator-and-authorizer.md]: This is just kind of continuation of the previous sentence. Basically I just meant that access to `/api`, `/api/*`, `/apis`, `/apis/*`and `/version` is essential and access to /metrics`, `/resetMetrics`, `/logs`, `/debug`, `/healthz`, `/swagger-ui/`, `/swaggerapi/`, `/ui` is optional.

**Comment by avlasov-mos-de** [comment]
Doesn't look like errors related to changes. Any suggestion on how to proceed?

**Comment by avlasov-mos-de** [comment]
it is still failing with not related error. I would appreciate any help and ideas on what further steps might be. I'm ready to learn how test framework for this and debug test itself if this can help(so any links are welcome in this case).

**Comment by ramineni** [comment]
@avlasov-mos-de yes, e2e conformance job failing on all PRs .. Already posted the same to openlab team to look into the issue ..hopefully should be resolved soon.

**Comment by avlasov-mos-de** [comment]
I wonder if  it is possible to proceed with review without this test? As it looks like it is not going to be fixed and I don't how to help with that.

**Comment by lingxiankong** [comment]
we've been meeting with some issue in the CI, the openlab team is working on this.

PR looks good to me.

/lgtm

**Comment by avlasov-mos-de** [comment]
Thanks @lingxiankong, does it mean it still has to pass that failing test before get 'approved ' label?

**Comment by lingxiankong** [comment]
@avlasov-mos-de no, technically speaking, we could merge PR even the CI fails, I'm not against that, considering this PR has nothing to do with the failed jobs. But i'd like to see other core reviewers' opinion for merge it or not.

**Comment by avlasov-mos-de** [comment]
So dear @reviewres/@approvers any chances we can move forward with this?

**Comment by dims** [comment]
@avlasov-mos-de looks like you may have to rebase

**Comment by avlasov-mos-de** [comment]
From what I can see here https://prow.k8s.io/pr?query=is%3Apr+repo%3Akubernetes%2Fcloud-provider-openstack+author%3Aavlasov-mos-de+head%3Amaster looks like the is now conflicts and all labels are there, however all comes to this broken test

**Pull request merged**